src/core/ddsc/src/dds_read.c

```diff
@@ -509,32 +510,36 @@ dds_return_t dds_take_next_wl (dds_entity_t reader, void **buf, dds_sample_info_
   return dds_read_impl (true, reader, buf, 1u, 1u, si, mask, DDS_HANDLE_NIL, true, true);
 }
 
-dds_return_t dds_return_loan (dds_entity_t reader_or_condition, void **buf, int32_t bufsz)
+dds_return_t dds_return_loan (dds_entity_t entity, void **buf, int32_t bufsz)
 {
   dds_reader *rd;
-  dds_entity *entity;
+  dds_entity *p_entity;
   dds_return_t ret;
 
   if (buf == NULL || (buf[0] == NULL && bufsz > 0) || (buf[0] != NULL && bufsz <= 0))
     return DDS_RETCODE_BAD_PARAMETER;
 
-  if ((ret = dds_entity_pin (reader_or_condition, &entity)) < 0) {
+  if ((ret = dds_entity_pin (entity, &p_entity)) < 0) {
     return ret;
-  } else if (dds_entity_kind (entity) == DDS_KIND_READER) {
-    rd = (dds_reader *) entity;
-  } else if (dds_entity_kind (entity) != DDS_KIND_COND_READ && dds_entity_kind (entity) != DDS_KIND_COND_QUERY) {
-    dds_entity_unpin (entity);
+  } else if (dds_entity_kind (p_entity) == DDS_KIND_READER) {
+    rd = (dds_reader *) p_entity;
+  } else if (dds_entity_kind (p_entity) == DDS_KIND_WRITER) {
+    ret = dds_return_writer_loan((dds_writer *) p_entity, *buf);
```

Comment:
`*buf` might be invalid here — something surprising enough that I felt it necessary to add a comment in the `bufsz <= 0`. (I do remember that my rationale was that this allows you to call `dds_return_loan` after a failed read/take operation simply passing in the latters result.)